sr_speech_control/src/sr_speech_control/speech_control.py

```diff
@@ -15,34 +15,56 @@
 from difflib import get_close_matches
 from std_msgs.msg import String
+import yaml
+from builtins import input
 
 
 class SpeechControl(object):
-    def __init__(self):
-        self.microphone = sr.Microphone()
-        prefer_microphone = rospy.get_param('~prefer_microphone')
-        if prefer_microphone:
-            for i, microphone_name in enumerate(sr.Microphone.list_microphone_names()):
-                if prefer_microphone in microphone_name:
-                    self.microphone = sr.Microphone(device_index=i)
-                    rospy.loginfo("Using preferred microphone: {}".format(microphone_name))
-                    break
-        self.trigger_word = rospy.get_param('~trigger_word', 'shadow')
+    def __init__(self, trigger_word, command_words, command_topic='sr_speech_control',
+                 similar_words_dict_path=None, non_speaking_duration=0.2, pause_threshold=0.2):
+        self.trigger_word = trigger_word
+        self.command_words = command_words
         self.recognizer = sr.Recognizer()
-        self._set_param_if_provided(self.recognizer, 'non_speaking_duration')
-        self._set_param_if_provided(self.recognizer, 'pause_threshold')
-        self.command_words = rospy.get_param('~command_words', [])
-        topic = rospy.get_param('~topic', 'sr_speech_control')
-        self.command_publisher = rospy.Publisher(topic, String, queue_size=1)
+        self.command_publisher = rospy.Publisher(command_topic, String, queue_size=1)
+        self.command_to_be_executed = None
+        self.similar_words_dict = {}
+
+        if similar_words_dict_path:
+            self.parse_similar_words_dict(similar_words_dict_path)
+        self._init_recognizer(non_speaking_duration, pause_threshold)
         self._stop_listening = self.recognizer.listen_in_background(self.microphone, self._recognizer_callback)
 
-    def _set_param_if_provided(self, object_to_set, param_name):
-        if rospy.has_param('~' + param_name):
-            setattr(object_to_set, param_name, rospy.get_param('~' + param_name))
+    def parse_similar_words_dict(self, path_name):
+        with open(path_name, 'r') as stream:
+            self.similar_words_dict = yaml.safe_load(stream)
+
+    def _init_recognizer(self, non_speaking_duration, pause_threshold):
+        for idx, mic in enumerate(sr.Microphone.list_microphone_names()):
+            rospy.loginfo('{}: {}'.format(idx, mic))
+
+        while not rospy.is_shutdown():
+            try:
+                idx = input("Choose one of the microphones from the list above. "
+                            "Type the index or leave empty for default microphone, than press [RETURN]\n")
+                if not idx:
+                    self.microphone = sr.Microphone()
+                else:
+                    self.microphone = sr.Microphone(device_index=int(idx))
+                with self.microphone as source:
+                    self.recognizer.adjust_for_ambient_noise(source)
```

Comment:
shouldn't the try statement be before the with instead of enclosing also the idx stuff?

Reply:
No, this is specifically there so there so that it checks is the microphone is valid within the try. What about the idx stuff?